mkosi/config.py

```diff
@@ -4295,3 +4340,87 @@ def systemd_tool_version(config: Config, tool: PathString) -> GenericVersion:
             sandbox=config.sandbox(binary=tool),
         ).stdout.split()[2].strip("()").removeprefix("v")
     )
+
+
+@dataclasses.dataclass(frozen=True)
+class CompletionItem:
+    short: Optional[str]
+    long: Optional[str]
+    help: Optional[str]
+    nargs: Union[str, int]
+    choices: list[str]
+    compgen: CompGen
+
+
+def collect_completion_arguments() -> list[CompletionItem]:
+    parser = create_argument_only_parser()
+
+    options = [
+        CompletionItem(
+            short=next((s for s in action.option_strings if not s.startswith("--")), None),
+            long=next((s for s in action.option_strings if s.startswith("--")), None),
+            help=action.help,
+            nargs=action.nargs or 0,
+            choices=[str(c) for c in action.choices] if action.choices is not None else [],
+            compgen=CompGen.from_action(action),
+        )
+        for action in parser._actions
+        if action.option_strings and action.help != argparse.SUPPRESS
+    ]
+
+    options += [
+        CompletionItem(
+            short=setting.short,
+            long=setting.long,
+            help=setting.help,
+            nargs=setting.nargs or 1,
+            choices=[str(c) for c in setting.choices] if setting.choices is not None else [],
+            compgen=setting.compgen,
+        )
+        for setting in SETTINGS
+    ]
+
+    return options
+
+
+def finalize_completion_bash(options: list[CompletionItem], resources: Path) -> str:
+    def to_bash_array(name: str, entries: Iterable[str]) -> str:
+        return f"declare -a {name.replace('-', '_')}=(" + " ".join(shlex.quote(str(e)) for e in entries) + ")"
+
+    def to_bash_hasharray(name: str, entries: Mapping[str, Union[str, int]]) -> str:
+        return (
+            f"declare -A {name.replace('-', '_')}=(" +
+            " ".join(f"[{shlex.quote(str(k))}]={shlex.quote(str(v))}" for k, v in entries.items()) + ")"
+        )
+
+    completion = resources / "completion.bash"
+
+    options_by_key = {o.short: o for o in options if o.short} | {o.long: o for o in options if o.long}
+
+    with io.StringIO() as c:
+        c.write(to_bash_array("_mkosi_options", options_by_key.keys()))
```

Comment:
Hmm, why `io.StringIO`? Why not just use `print()` here?

Reply:
Probably premature optimization. The idea is, that distro packages ship the already built completion script, but in case somebody really wants to generate it themselves, `io.StringIO()` is the fastest way to to generate a string. Also, this way the whole output gets built or not, it can't be generated and printed half. At least that was the idea.

Reply:
I guess not writing the partial string makes sense. But can it really fail?

At least use `print(…, file=out)` instead of `c.write()`.